include/fmt/format.h

```diff
@@ -1822,85 +1843,201 @@ class specs_handler: public specs_setter<typename Context::char_type> {
+  arg_ref_type width_ref;
+  arg_ref_type precision_ref;
 };
 
 // Format spec handler that saves references to arguments representing dynamic
 // width and precision to be resolved at formatting time.
-template <typename ParseContext>
-class dynamic_specs_handler :
-    public specs_setter<typename ParseContext::char_type> {
+template <typename Specs, typename ParseContext, typename ArgRefCreator>
+class dynamic_specs_handler
+    : public specs_setter<typename ParseContext::char_type> {
  public:
   typedef typename ParseContext::char_type char_type;
 
-  FMT_CONSTEXPR dynamic_specs_handler(
-      dynamic_format_specs<char_type> &specs, ParseContext &ctx)
-    : specs_setter<char_type>(specs), specs_(specs), context_(ctx) {}
+  FMT_CONSTEXPR dynamic_specs_handler(Specs &specs, ParseContext &ctx,
+                                      ArgRefCreator &creator)
+      : specs_setter<char_type>(specs),
+        specs_(specs),
+        context_(ctx),
+        arg_ref_creator_(creator) {}
 
   FMT_CONSTEXPR dynamic_specs_handler(const dynamic_specs_handler &other)
-    : specs_setter<char_type>(other),
-      specs_(other.specs_), context_(other.context_) {}
+      : specs_setter<char_type>(other),
+        specs_(other.specs_),
+        context_(other.context_),
+        arg_ref_creator_(other.arg_ref_creator_) {}
 
   template <typename Id>
   FMT_CONSTEXPR void on_dynamic_width(Id arg_id) {
-    specs_.width_ref = make_arg_ref(arg_id);
+    context_.check_arg_id(arg_id);
+    specs_.width_ref = arg_ref_creator_.create(arg_id);
   }
 
   template <typename Id>
   FMT_CONSTEXPR void on_dynamic_precision(Id arg_id) {
-    specs_.precision_ref = make_arg_ref(arg_id);
+    context_.check_arg_id(arg_id);
+    specs_.precision_ref = arg_ref_creator_.create(arg_id);
   }
 
   FMT_CONSTEXPR void on_error(const char *message) {
     context_.on_error(message);
   }
 
  private:
-  typedef arg_ref<char_type> arg_ref_type;
+  Specs &specs_;
+  ParseContext &context_;
+  ArgRefCreator &arg_ref_creator_;
+};
 
-  template <typename Id>
-  FMT_CONSTEXPR arg_ref_type make_arg_ref(Id arg_id) {
-    context_.check_arg_id(arg_id);
-    return arg_ref_type(arg_id);
+template <typename ParseContext>
+class dynamic_arg_ref_creator {
+  typedef typename ParseContext::char_type char_type;
+
+ public:
+  typedef arg_ref<char_type, string_value<char_type>> arg_ref_type;
+
+  FMT_CONSTEXPR dynamic_arg_ref_creator(ParseContext &ctx) : context_(ctx) {}
+
+  FMT_CONSTEXPR arg_ref_type create(unsigned id) { return arg_ref_type(id); }
+
+  FMT_CONSTEXPR arg_ref_type create(basic_string_view<char_type> id) {
+    return arg_ref_type(string_value<char_type>{id.data(), id.size()});
   }
 
-  FMT_CONSTEXPR arg_ref_type make_arg_ref(auto_id) {
+  FMT_CONSTEXPR arg_ref_type create(auto_id) {
     return arg_ref_type(context_.next_arg_id());
   }
 
-  dynamic_format_specs<char_type> &specs_;
+ private:
   ParseContext &context_;
 };
 
+struct string_view_metadata {
```

Comment:
Why do we need this class? Can it be replaced with `string_view`?

Reply:
Prepared format needs to be copyable and movable. User can store it as a class member, in containers etc. 
Prepared format stores a copy of the format string, from which it has been created. Beside the format, it stores format parts, and format part can store a 'view' to a part of the primary format string, e.g. to store a name of an argument, or so.
Now consider such situation with usage of `string_view`:
1. User creates a prepared format, e.g. `"{arg}"`. Let's call it `PreparedFormat_1`.
1.1 Prepared format stores a copy of the format string as a `std::basic_string<>("{arg"})`. Let's call it `FormatString_1`.
1.2 Prepared format creates format parts. In fact one part with a 'view' that is a `string_view`, which represents a view of string `"arg"` that is a part of the above `FormatString_1`.
2. User copies `PreparedFormat_1` to a `PreparedFormat_2`.
2.1 `PreparedFormat_2` creates a copy of the `FormatString_1`. Let's call it `FormatString_2`.
2.2 `PreparedFormat_2` creates a copy of the `PreparedFormat_1`'s format parts.
3. User destroys `PreparedFormat_1`.
3.1 `FormatString_1` and `PreparedFormat_1`'s format parts are destroyed.
4. User ends up with a `PreparedFormat_2`, but `PreparedFormat_2`'s format parts are corrupted. The part that represents a name of the argument "arg" is a `string_view` that refers to the destroyed `FormatString_1`, not the `FormatString_2`.

That's why the `string_view_metadata` was introduced. It's representation of a string view that is not bind to any actual string. Later on, when the view is needed, and we have a primary string, we can create a `string_view`, that refers to that primary string. It's done in the `prepared_format` class, where it actually needs the `string_view`s.

An alternative would be to store views as `string_view`s, and recreate them (in fact, not only them but all the prepared parts) on copy/move of a prepared format. I decided to not go this way because IMO it's more error-prone and complicated than the `string_view_metadata` way.

Reply:
>Why do we need unrestricted union

TBH I don't know about which union are you asking. Could you please elaborate?

Reply:
`arg_ref::value`

Reply:
So, do you suggest to simply use `struct` there? Why? `union` can be used with almost zero cost. Yes, we need to type more code, but couple of bytes are saved with every `arg_ref`. `arg_ref`s are used in prepared format parts, so we're also saving couple user's memory bytes. Let me know if that answers your concerns. I'm still not sure if I got it right.

Reply:
> So, do you suggest to simply use struct there?

No, I'm suggesting to use a normal (not unrestricted) union and don't fallback to `struct`. This is a minor point though and can be addressed later.

Reply:
I see. I think that's not possible (I mean it is, but it'd be more problematic).
1. Old compilers need restricted union - union without any ctor etc.
2. Modern compilers need constexpr union - union with at least one constexpr ctor (in this case, we need those three ctors).

There is a way that I can think of - create two separate unions and ifdef them for old and modern compiler. I decided that it's more problematic and using the `FMT_UNRESTRICTED_UNION` is actually a cleaner way. What do you think?

Reply:
Fair enough. Let's keep `FMT_UNRESTRICTED_UNION` here.